docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,298 @@
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU corev1.ResourceName = "cpu"
+    AutoscalerResourceMemory corev1.ResourceName = "memory"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity corev1.ResourceList `json:"capacity,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  machine:
+    capacity:
+      memory: 500mb
+      cpu: "1"
+      nvidia.com/gpu: "1"
+```
+
+#### MachineSet and MachineDeployment Annotations
+
+In cases where a user needs to provide specific resource information for a
+MachineSet or MachineDeployment, or in cases where an infrastructure provider
+has not yet added the Infrastructure Machine Template status changes, they
+may use annotations to provide the information. The annotation values match the
+API that is used in the Infrastructure Machine Template, e.g. the memory and cpu
+annotations allow a `resource.Quantity` value and the two gpu annotations allow
+for the count and type of GPUs per instance.
+
+If a user wishes to specify the resource capacity through annotations, they
+may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
+that are participating in autoscaling:
+
+```
+kind: <MachineSet or MachineDeployment>
+metadata:
+  annotations:
+      capacity.cluster.x-k8s.io/gpu-count: "1"
+      capacity.cluster.x-k8s.io/gpu-type: "nvidia.com/gpu"
+      capacity.cluster.x-k8s.io/memory: "500mb"
+      capacity.cluster.x-k8s.io/cpu: "1"
```

Comment:
Following up from yesterday's community meeting, maybe these annotations could be defined in Autoscaler directly and supported there?

Reply:
In what sense? I think this was always going to be on the CA side in some way.

Reply:
Oh, as in not have them live in the `cluster.x-k8s.io` namespace?

Reply:
i was going to remove the annotations from this enhancement since it sounded like we wanted to not have another point for the api interaction. do we want to keep them?

Reply:
Up to you, my worry was mostly around exposing these annotations in Cluster API, and under the cluster.x-k8s.io domain

Reply:
ok, let me do a little research on the autoscaler side. if it makes sense then we could keep them with something like `capacity.autoscaler.x-k8s.io` ?

Reply:
Works for me if those variables are only exposed on the autoscaler side

Reply:
i'm not fully following, what do you mean by "only exposed on the autoscaler side" ?

Reply:
I think that the constants are added to the autoscaler codebase and not to CAPI.